File: src/templates/python/agent_basic.py
"""
Token Optimisation Proxy — Python Developer Starter Kit (Basic)

Replace PROXY_ENDPOINT and PROXY_API_KEY with values from your platform team.
Do NOT use LLM provider keys (OpenAI, Anthropic, etc.) — the proxy handles all of that.

Usage:
    export PROXY_API_KEY=<your-proxy-key>
    export PROXY_ENDPOINT=https://token-proxy-<hash>-uc.a.run.app
    python agent_basic.py
"""
import os
from openai import OpenAI

PROXY_ENDPOINT = os.environ["PROXY_ENDPOINT"]   # provided by platform team
PROXY_API_KEY  = os.environ["PROXY_API_KEY"]    # provided by platform team — NOT an OpenAI key

client = OpenAI(
    api_key=PROXY_API_KEY,
    base_url=f"{PROXY_ENDPOINT}/v1",
)


def ask(prompt: str, model: str = "gpt-4o-mini") -> str:
    response = client.chat.completions.create(
        model=model,
        messages=[{"role": "user", "content": prompt}],
        max_tokens=512,
        extra_body={
            "user": os.environ.get("USER", "dev"),  # for per-user savings tracking
        },
    )
    return response.choices[0].message.content


def ask_with_session(prompt: str, session_id: str, model: str = "gpt-4o-mini") -> str:
    """Multi-turn conversation — G10 memory management applied automatically."""
    response = client.chat.completions.create(
        model=model,
        messages=[{"role": "user", "content": prompt}],
        max_tokens=512,
        extra_body={
            "user": os.environ.get("USER", "dev"),
            "x_session_id": session_id,  # enables G10 sliding window memory
        },
    )
    savings = response.model_extra.get("_token_opt", {})
    print(f"  Saved: {savings.get('total_abs_saving', 0)} tokens "
          f"({savings.get('total_pct_saving', 0):.1f}%)")
    return response.choices[0].message.content


if __name__ == "__main__":
    print(ask("What is the capital of France?"))
    print(ask_with_session("Summarise the history of Python.", session_id="demo-session-1"))

File: src/templates/java/AgentBasic.java
package com.example.tokenopt;

import com.openai.client.OpenAIClient;
import com.openai.client.okhttp.OpenAIOkHttpClient;
import com.openai.models.ChatCompletion;
import com.openai.models.ChatCompletionCreateParams;
import com.openai.models.ChatCompletionMessageParam;
import com.openai.models.ChatCompletionUserMessageParam;

import java.util.List;
import java.util.Map;

/**
 * Token Optimisation Proxy — Java Developer Starter Kit
 *
 * Replace PROXY_ENDPOINT and PROXY_API_KEY env vars with values from your platform team.
 * Do NOT use LLM provider keys — the proxy handles all provider authentication.
 *
 * Maven dependency: com.openai:openai-java:0.9.0+
 */
public class AgentBasic {

    private static final String PROXY_ENDPOINT = System.getenv("PROXY_ENDPOINT");
    private static final String PROXY_API_KEY   = System.getenv("PROXY_API_KEY");

    private final OpenAIClient client;

    public AgentBasic() {
        this.client = OpenAIOkHttpClient.builder()
            .apiKey(PROXY_API_KEY)
            .baseUrl(PROXY_ENDPOINT + "/v1")
            .build();
    }

    public String ask(String prompt) {
        ChatCompletionCreateParams params = ChatCompletionCreateParams.builder()
            .model("gpt-4o-mini")
            .messages(List.of(
                ChatCompletionMessageParam.ofUser(
                    ChatCompletionUserMessageParam.builder()
                        .content(prompt)
                        .build()
                )
            ))
            .maxTokens(512)
            .build();

        ChatCompletion completion = client.chat().completions().create(params);
        return completion.choices().get(0).message().content().orElse("");
    }

    public String askWithSession(String prompt, String sessionId) {
        ChatCompletionCreateParams params = ChatCompletionCreateParams.builder()
            .model("gpt-4o-mini")
            .messages(List.of(
                ChatCompletionMessageParam.ofUser(
                    ChatCompletionUserMessageParam.builder()
                        .content(prompt)
                        .build()
                )
            ))
            .maxTokens(512)
            // G10: session-aware memory management
            .putAdditionalBodyProperty("x_session_id", sessionId)
            // G17: workflow budget tracking
            .putAdditionalBodyProperty("workflow_id", "java-workflow-" + sessionId)
            .putAdditionalBodyProperty("user", System.getProperty("user.name", "java-dev"))
            .build();

        ChatCompletion completion = client.chat().completions().create(params);
        return completion.choices().get(0).message().content().orElse("");
    }

    public static void main(String[] args) {
        AgentBasic agent = new AgentBasic();
        System.out.println(agent.ask("What is machine learning?"));
        System.out.println(agent.askWithSession("Summarise recent AI trends.", "session-001"));
    }
}
